### PR TITLE
Improvement/#34 argparse and config file for simulation

### PR DESCRIPTION
**Pull request by JAfzali**

### CODE REVIEW

**Comment by NoB0** [comment]
- Introduce a default config file to run the simulation
- Load and update configuration with command line arguments
- Save simulation configuration in `data/runs` 
- Load intent classifier supported by DialogueKit (i.e., cosine and Rasa DIET)
- Add abstract methods in `SampleAgent` to run `run_simulation.py`
- Create #88 as follow up to mypy issue

**Comment by kbalog** [comment]
@NoB0 Is this ready to be reviewed or are you still working on it?

**Comment by NoB0** [comment]
It can be reviewed.

**Review by kbalog — CHANGES_REQUESTED** [review]
Wondering if `get_intent_classifier()`, `load_cosine_classifier()`, and `load_rasa_diet_classifier()` could be moved to DialogueKit instead?  Or at least the latter two? They could receive specific arguments (intent schema, annotated dialogues file, etc.) instead of a config file.

File: config/default/config_default.yaml [hunk 1]
```diff
@@ -0,0 +1,9 @@
+agent_id: "SampleAgent"
+output_name: "sample_agent"
+ontology: data/domains/movies.yaml
```
[the remark this answers is not in the document]
Comment (NoB0) [reply config/default/config_default.yaml:3]: Done.

File: config/default/config_default.yaml [hunk 2]
```diff
@@ -0,0 +1,9 @@
...
+intents: data/interaction_models/crs_v1.yaml
+items: data/movielens-25m-sample/movies_w_keywords.csv
+ratings: data/movielens-25m-sample/ratings.csv
+dialogues: data/agents/moviebot/annotated_dialogues.json
+intent_classifier: "cosine"
+rasa_dialogues: data/agents/moviebot/annotated_dialogues_rasa_agent.yml
```
[the remark this answers is not in the document]
Comment (NoB0) [reply config/default/config_default.yaml:9]: Done.

File: usersimcrs/run_simulation.py [hunk 3]
```diff
@@ -24,85 +30,66 @@
...
-    if not os.path.exists(args.ratings):
-        sys.exit("FileNotFound: {file}".format(file=args.ratings))
+    Args:
+        config: Configuration generated from YAML configuration file.
+    """
+    agent = SampleAgent(agent_id="Tester")
 
-    domain = Domain(args.domain)
+    # Loads domain, item collection, and preference data
+    domain = Domain(config["ontology"].get())
```
[the remark this answers is not in the document]
Comment (NoB0) [reply usersimcrs/run_simulation.py:54]: Done.

File: usersimcrs/run_simulation.py [hunk 4]
```diff
@@ -112,3 +99,195 @@ def simulate_conversation(
...
+    """Loads config from config file and command line parameters.
+
+    Loads default values from `config/default/config_default.yaml`. Values are
+    then updated with any value specified in the command line arguments.
+
+    Args:
+        args: Arguments parsed with argparse.
+    """
+    # Load default config
+    config = confuse.Configuration("usrsimcrs")
```
[the remark this answers is not in the document]
Comment (NoB0) [reply usersimcrs/run_simulation.py:185]: Fixed.

File: usersimcrs/run_simulation.py [hunk 5]
```diff
@@ -112,3 +99,195 @@ def simulate_conversation(
...
+    config.set_file(DEFAULT_CONFIG_PATH.format("general"))
```
[the remark this answers is not in the document]
Comment (NoB0) [reply usersimcrs/run_simulation.py:186]: Removed.

File: usersimcrs/run_simulation.py [hunk 6]
```diff
@@ -112,3 +99,195 @@ def simulate_conversation(
...
+
+    # Load simulation specific config
+    agent_id = args.agent_id or config["agent_id"].get()
+    config.set_file(DEFAULT_CONFIG_PATH.format(agent_id))
```
[the remark this answers is not in the document]
Comment (NoB0) [reply usersimcrs/run_simulation.py:190]: Removed.

File: usersimcrs/run_simulation.py [hunk 7]
```diff
@@ -112,3 +99,195 @@ def simulate_conversation(
...
+    # Load additional config (update defaults).
+    if args.config_file:
+        config.set_file(args.config_file)
+
+    # Update config from command line arguments
+    config.set_args(args, dots=True)
+
+    # Save run config to metadata file
+    output_name = config["output_name"].get()
+    with open(f"data/runs/{output_name}.meta.yaml", "w") as f:
```
[the remark this answers is not in the document]
Comment (NoB0) [reply usersimcrs/run_simulation.py:201]: Done.

**Comment by NoB0** [comment]
`get_intent_classifier` is specific to this repository and should not be moved in my opinion. 
I suppose the methods `load_cosine_classifier` and `load_rasa_diet_classifier` could be "utils" function in DK.

**Pull request merged**